index.bs

```diff
@@ -1899,23 +1905,27 @@ The <dfn method for="WebGLRenderingContextBase">makeXRCompatible()</dfn> method
 When this method is invoked, the user agent MUST run the following steps:
 
   1. Let |promise| be [=a new Promise=].
-    1. Run the following steps [=in parallel=]:
-    1. Let |context| be the target {{WebGLRenderingContextBase}} object.
-    1. If |context|'s [=WebGL context lost flag=] is set, [=reject=] |promise| with an {{InvalidStateError}} and abort these steps.
-    1. If |context|'s [=XR compatible=] boolean is <code>true</code>, [=/resolve=] |promise| and abort these steps.
-    1. [=ensures an immersive XR device is selected|Ensure an immersive XR device is selected=].
-    1. If the [=XR/immersive XR device=] is <code>null</code>:
-        1. Set |context|'s [=XR compatible=] boolean to <code>false</code>.
-        1. [=Reject=] |promise| with an {{InvalidStateError}} and abort these steps.
-    1. If |context| was created on a [=compatible graphics adapter=] for the [=XR/immersive XR device=]:
-        1. Set |context|'s [=XR compatible=] boolean to <code>true</code>.
-        1. [=/Resolve=] |promise| and abort these steps.
-    1. [=Queue a task=] to perform the following steps:
-        1. Force |context| to be lost and [=handle the context loss=] as described by the WebGL specification.
-        1. If the [=canceled flag=] of the "webglcontextlost" event fired in the previous step was not set, [=reject=] |promise| with an {{AbortError}} and abort these steps.
-        1. [=Restore the context=] on a [=compatible graphics adapter=] for the [=XR/immersive XR device=].
-        1. Set |context|'s [=XR compatible=] boolean to <code>true</code>.
-        1. [=/Resolve=] |promise|.
+  1. Let |context| be the target {{WebGLRenderingContextBase}} object.
+  1. [=ensures an immersive XR device is selected|Ensure an immersive XR device is selected=].
+  1. Set |context|'s [=XR compatible=] boolean as follows:
+    <dl class="switch">
+      <dt> If |context|'s [=WebGL context lost flag=] is set
+      <dd> Set |context|'s [=XR compatible=] boolean to <code>false</code> and [=reject=] |promise| with an {{InvalidStateError}}.
+      <dt> If the [=XR/immersive XR device=] is <code>null</code>
+      <dd> Set |context|'s [=XR compatible=] boolean to <code>false</code> and [=reject=] |promise| with an {{InvalidStateError}}.
+      <dt> If |context|'s [=XR compatible=] boolean is <code>true</code>
+      <dd> [=/Resolve=] |promise|.
+      <dt> If |context| was created on a [=compatible graphics adapter=] for the [=XR/immersive XR device=]
+      <dd> Set |context|'s [=XR compatible=] boolean to <code>true</code> and [=/resolve=] |promise|.
+      <dt> Otherwise
+      <dd> [=Queue a task=] to perform the following steps:
```

Comment:
It'd be better to say "Queue a task on the WebGL task source to ...", I think, instead of putting the task source part completely separate from this operation.

Also, this still needs to define the document for this task...

Reply:
Ah, interesting. I was primarily mimicking the pattern that I saw repeated throughout the HTML standard. For example:
 - https://html.spec.whatwg.org/multipage/canvas.html#dom-canvas-toblob
 - https://html.spec.whatwg.org/multipage/interactive-elements.html#details-notification-task-steps
 - https://html.spec.whatwg.org/multipage/browsing-the-web.html#scroll-to-fragid:task-source

Happy to change it, though. (Also none of those examples define a document, unless I'm overlooking something, so it wasn't clear what was required in that case.)

Reply:
The HTML standard has large parts that predate the "modern" formulation of queueing tasks, and also some sections where lots of tasks get queued and it tries to coalesce the "which task source" verbiage.

In practice, this is really annoying to deal with and makes it hard to find what task source to use for tasks, and I wish the HTML standard would stop doing that.

And yes, most things don't documents at the moment, but that's being worked on.  See https://github.com/whatwg/html/issues/4980

Reply:
Thanks for the explanation! If you don't mind providing a bit more guidance, I've got a followup question.

In this specific case it seems like the appropriate document is the document which created the canvas element which created the WebGL context, unless the context was created by an OffscreenCanvas, in which case I'm not sure how to reason about what the appropriate document is? (It may have been created in a worker.)

Reply:
Right, if we have an element it's easy.  If we don't, then I don't know enough about the relevant APIs to tell you.  What does Chrome implement, since they actually do document-association of tasks?